pkg/provisioner/crdprovisioner.go

```diff
@@ -0,0 +1,265 @@
+		azDriverNodeUpdate = azDriverNodeFromCache.DeepCopy()
+	} else if errors.IsNotFound(err) {
+		// If AzDriverNode object is not there create it
+		klog.Errorf("AzDriverNode is not registered yet, will create. error: %v", err)
+		azDriverNodeNew := &v1alpha1.AzDriverNode{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: nodeID,
+			},
+			Spec: v1alpha1.AzDriverNodeSpec{
+				NodeName: nodeID,
+			},
+		}
+		if azDriverNodeNew.Labels == nil {
+			azDriverNodeNew.Labels = make(map[string]string)
+		}
+		azDriverNodeNew.Labels[azureutils.PartitionLabel] = nodePartition
+		klog.V(2).Infof("Creating AzDriverNode with details (%v)", azDriverNodeNew)
+		azDriverNodeCreated, err := azN.Create(ctx, azDriverNodeNew, metav1.CreateOptions{})
+		if err != nil || azDriverNodeCreated == nil {
+			klog.Errorf("Failed to create/update azdrivernode resource for node (%s), error: %v", nodeID, err)
+			return err
+		}
+		azDriverNodeUpdate = azDriverNodeCreated.DeepCopy()
+	} else {
+		klog.Errorf("Failed to get AzDriverNode for node (%s), error: %v", nodeID, err)
+		return errors.NewBadRequest("Failed to get AzDriverNode or node not found, can not register the plugin.")
+	}
+
+	// Do an initial update to AzDriverNode status
+	if azDriverNodeUpdate.Status == nil {
+		azDriverNodeUpdate.Status = &v1alpha1.AzDriverNodeStatus{}
+	}
+	readyForAllocation := false
+	timestamp := time.Now().UnixNano()
+	statusMessage := "Driver node initializing."
+	azDriverNodeUpdate.Status.ReadyForVolumeAllocation = &readyForAllocation
+	azDriverNodeUpdate.Status.LastHeartbeatTime = &timestamp
+	azDriverNodeUpdate.Status.StatusMessage = &statusMessage
+	klog.V(2).Infof("Updating status for AzDriverNode Status=(%v)", azDriverNodeUpdate)
+	_, err = azN.UpdateStatus(ctx, azDriverNodeUpdate, metav1.UpdateOptions{})
+	if err != nil {
+		klog.Errorf("Failed to update status of azdrivernode resource for node (%s), error: %v", nodeID, err)
+		return err
+	}
+
+	return nil
+}
+
+func (c *CrdProvisioner) CreateVolume(
+	ctx context.Context,
+	volumeName string,
+	capacityRange *v1alpha1.CapacityRange,
+	volumeCapabilities []*v1alpha1.VolumeCapability,
+	parameters map[string]string,
+	secrets map[string]string,
+	volumeContentSource *v1alpha1.ContentVolumeSource,
+	accessibilityReq *v1alpha1.TopologyRequirement) (*v1alpha1.AzVolumeStatusParams, error) {
+	azV := c.azDiskClient.DiskV1alpha1().AzVolumes(c.namespace)
+	azVolume := &v1alpha1.AzVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: volumeName,
+		},
+		Spec: v1alpha1.AzVolumeSpec{
+			UnderlyingVolume:          volumeName,
+			VolumeCapability:          volumeCapabilities,
+			CapacityRange:             capacityRange,
+			Parameters:                parameters,
+			Secrets:                   secrets,
+			ContentVolumeSource:       volumeContentSource,
+			AccessibilityRequirements: accessibilityReq,
+		},
+	}
+
+	azVolumeCreated, err := azV.Create(ctx, azVolume, metav1.CreateOptions{})
+	if err != nil || azVolumeCreated == nil {
```

Comment:
I wouldn't expect `azVolumeCreated` to be `nil` if `err == nil`, but if this truly is possible, it needs to be handled separately. Otherwise the caller will receive `nil, nil` and crash.